WIP: Enhanced Deletion

samdeane opened the pull request:
Work in progress on some changes for use in Agenda.

Adds the facility to replace deleted blob files with a "tombstone" marker file which forces the rest of the API to treat the blob as deleted even if the actual data file returns.

This is intended to work around situations where the blob directory is being synced with an external service that might get confused and put back deleted files instead of removing them.

*Note*: the behaviour of the existing `deleteBlobAtPath` call should be unaffected by these changes. A new variant of the call has been added to opt-in to tombstones. Other API has been updated to be aware of the presence of tombstone files, but should be unaffected if they are never created. There will be a minor reduction in efficiency due to the need to check for tombstones, but it shouldn't be significant in comparison to whatever work is actually done with the files.

samdeane committed:
Started stubbing tombstone support.

samdeane committed:
Added blob deletion tests

samdeane committed:
Improved deletion tests

samdeane committed:
Added constant for tombstone extension

samdeane committed:
Don't return data and don't enumerate files where a tombstone is present.

samdeane committed:
Renamed tests.

samdeane committed:
comment tweak

samdeane committed:
Added blobExistsAtPath

samdeane committed:
Honour the usingTombstone flag.

samdeane committed:
Added comments about potential cleanup.

samdeane committed:
Added a note about double-deletion.

samdeane committed:
Added some comments explaining the tests

drewmccormack commented on Core/PARStore.h:
```diff
@@ -68,9 +68,11 @@ extern NSString *PARStoreDidSyncNotification;
 - (nullable NSString *)absolutePathForBlobPath:(NSString *)path;
 - (NSArray<NSString *> *)absolutePathsForBlobsPrefixedBy:(NSString *)prefix NS_SWIFT_NAME(absolutePaths(forBlobsPrefixedBy:));
 - (void)enumerateBlobs:(void(^)(NSString *path))block;
+- (BOOL)blobExistsAtPath:(NSString *)path;
```
I guess we would want to be able to query if a tombstone exists too. Imagine we are doing the sync engine. It would want to check if a particular file from the cloud is now a tombstone, and should be removed. Note that that is different to the blob simply not being found locally, which would indicate it should be downloaded, not deleted.

So I'm thinking probably want a query for tombstone existence too. Perhaps `blobIsRegisterdDeletedAtPath:`. It think this already implies the use of tombstones, otherwise how would we know.

samdeane replied:
Yeah, I wasn't quite sure whether we needed to treat 'file doesn't exist' as different from 'tombstone exists', or whether they'd turn out to imply the same thing in any calling code. That's what I was aiming at with just having the `exists` call, but we may well need the other one instead.

drewmccormack replied:
I think most of the time you just want the exists call, and it should handle tombstones for you. But in rare occasions, like sync, you want to query whether the formerly existed.

drewmccormack commented on Core/PARStore.h:
```diff
@@ -68,9 +68,11 @@ extern NSString *PARStoreDidSyncNotification;
 - (BOOL)writeBlobFromPath:(NSString *)sourcePath toPath:(NSString *)path error:(NSError **)error;
 - (nullable NSData *)blobDataAtPath:(NSString *)path error:(NSError **)error;
 - (BOOL)deleteBlobAtPath:(NSString *)path error:(NSError **)error;
+- (BOOL)deleteBlobAtPath:(NSString *)path usingTombstone: (BOOL)usingTombstone error:(NSError **)error;
```
Just wondering if we should leak the word "tombstone" into the API. In theory, that is purely an implementation detail.

How about `deleteBlobAtPath:registeringDeletion:error:`

samdeane replied:
Good plan.

drewmccormack commented on Core/PARStore.m:
```diff
@@ -1382,7 +1384,41 @@ - (BOOL)writeBlobFromPath:(NSString *)sourcePath toPath:(NSString *)targetSubpat
     return YES;
 }
 
+- (BOOL)writeTombstoneAtPath:(NSString *)tombstonePath forFileAtPath:(NSString *)filePath error:(NSError **)error
```
Just to be clear, although I suggest using "registeredDeletion" in the API, the private methods can stick with tombstone.

drewmccormack commented on Core/PARStore.m:
```diff
@@ -1407,15 +1443,36 @@ - (BOOL)deleteBlobAtPath:(NSString *)path error:(NSError **)error
+        // To be consistent with this behaviour, we should perhaps also return an error if we
+        // find an existing tombstone file.
+        // However, my normal inclination would be to treat the presence of the tombstone as a sign
+        // that the file has been successfully deleted and quietly return success.
```
I think it is safer to return an error. It gives the little bit more information. If you expect to find a file, and you don't, that is probably something you want to know about.

drewmccormack commented on Core/PARStore.m:
```diff
@@ -1407,15 +1443,36 @@ - (BOOL)deleteBlobAtPath:(NSString *)path error:(NSError **)error
+        NSError *error = nil;
+
+        BOOL success = !usingTombstone;
+        if (!success) {
```
Testing the success var here is a bit difficult to wrap my head around. Probably better to be explicit. What you are actually testing is if it is using tombstones. So test `usingTombstone`.

drewmccormack commented on Core/PARStore.m:
```diff
@@ -1407,15 +1443,36 @@ - (BOOL)deleteBlobAtPath:(NSString *)path error:(NSError **)error
+            // write tombstone
+            success = [self writeTombstoneAtPath:newTombstoneURL.path forFileAtPath:newURL.path error:&error];
+            if (!success)
+                localError = [NSError errorWithObject:self code:__LINE__ localizedDescription:[NSString stringWithFormat:@"Could not create tombstone for data blob at path '%@'", newURL.path] underlyingError:error];
```
Think I would prefer having the {} on the if, although it is long ago that I wrote ObjC.

samdeane replied:
I agree but I was trying to match the style of the other code. That line was copied & pasted from elsewhere in fact (then the error was changed).

drewmccormack replied:
OK, no problem. If that is the style of the rest, happy to keep it that way.

drewmccormack commented on Core/PARStore.m:
```diff
@@ -1407,15 +1443,36 @@ - (BOOL)deleteBlobAtPath:(NSString *)path error:(NSError **)error
+        if (success) {
+            success = [[NSFileManager defaultManager] removeItemAtURL:fileURL error:&error];
+            if (!success)
+                localError = [NSError errorWithObject:self code:__LINE__ localizedDescription:[NSString stringWithFormat:@"Could not delete data blob at path '%@'", newURL.path] underlyingError:error];
```
Same {} comment here.

drewmccormack commented on Core/PARStore.m:
```diff
@@ -1456,6 +1513,21 @@ - (NSData *)blobDataAtPath:(NSString *)path error:(NSError **)error;
+        NSString *description = [NSString stringWithFormat:@"Blob at path '%@' has a tombstone, indicating that it has been deleted.", self.storeURL.path];
+        ErrorLog(@"%@", description);
+        if (error != NULL)
+            *error = [NSError errorWithObject:self code:__LINE__ localizedDescription:description underlyingError:nil];
```
{}

drewmccormack commented on Core/PARStore.m:
```diff
@@ -1456,6 +1513,21 @@ - (NSData *)blobDataAtPath:(NSString *)path error:(NSError **)error;
+        if (error != NULL)
+            *error = [NSError errorWithObject:self code:__LINE__ localizedDescription:description underlyingError:nil];
+        
+        // TODO: should we attempt to clean up sync errors here, by deleting the data file if it still exists?
```
Yes, good idea. May as well remove the file if the tombstone exists. Probably some earlier crash.

drewmccormack reviewed (changes requested):
Looking good. Nice job!

I mentioned a few changes in the comments.

 There is one other I wanted to bring up: what happens if someone re-inserts a file that has a tombstone? I guess in our system, we don't want this, because if we remove the tombstone, we get the problem that a tombstone on other devices will still result in the file being deleted eventually. But perhaps what we can do is simply give an error if someone goes to insert a file that already has a tombstone.

samdeane committed:
Updated to reflect Drew's comments

samdeane committed:
Fixed blob enumeration. Improved tests. Added enumeration of tombstones.

drewmccormack commented on Core/PARStore.h:
```diff
@@ -68,12 +68,15 @@ extern NSString *PARStoreDidSyncNotification;
 - (BOOL)blobExistsAtPath:(NSString *)path;
+
+/// @name Registered Deletion Support
+- (BOOL)deleteBlobAtPath:(NSString *)path registeringDeletion: (BOOL)usingTombstone error:(NSError **)error;
```
Small thing, but variable name is still old one (usingTombstone)

drewmccormack commented on Core/PARStore.m:
```diff
@@ -1604,7 +1604,7 @@ - (void)enumerateBlobs:(void(^)(NSString *path))block
                 BOOL isDir = false;
                 if ([fileManager fileExistsAtPath:url.path isDirectory:&isDir] && !isDir) continue;
-                if (url.pathExtension == TombstoneFileExtension) {
+                if ([url.pathExtension isEqualToString: TombstoneFileExtension]) {
```
Close one :)

samdeane committed:
Pass out both the blob path and the tombstone path when enumerating tombstones.

samdeane committed:
Fixed old variable name.

samdeane committed:
Fixed enumeration - should be skipping directories.

samdeane committed:
Added option to ignore missing files when deleting.

We can't fill in all the tombstone metadata if the file has already been deleted, and arguably there is no point in making one anyway, so it's helpful to be able to choose to ignore this without failing.
